packages/react-juce/src/lib/BlueprintBackend.ts

```diff
@@ -146,30 +160,100 @@ export class RawTextViewInstance {
   }
 }
 
-export default {
+function __getRootContainer(): ViewInstance {
+  if (__rootViewInstance !== null)
+    return __rootViewInstance;
 
-  getRootContainer(): ViewInstance {
-    if (__rootViewInstance !== null)
-      return __rootViewInstance;
+  //@ts-ignore
+  const id = __BlueprintNative__.getRootInstanceId();
+  __rootViewInstance = new ViewInstance(id, 'View');
 
-    //@ts-ignore
-    const id = __BlueprintNative__.getRootInstanceId();
-    __rootViewInstance = new ViewInstance(id, 'View');
+  return __rootViewInstance;
+}
 
-    return __rootViewInstance;
-  },
+function __hasFunctionProp(view: ViewInstance, prop: string) {
+   return view._props.hasOwnProperty(prop) &&
+          typeof view._props[prop] === 'function';
+}
 
+function __callEventHandlerIfPresent(view: Instance, eventType: string, event: any) {
+  if (view instanceof ViewInstance && __hasFunctionProp(view, eventType)) {
+    view._props[eventType](event);
+  }
+}
+
+function __bubbleEvent(view: Instance, eventType: string, event: any): void {
+  if (view && view !== __getRootContainer()) {
+    // Always call the event callback on the target before bubbling.
+    // Some events may not bubble or have bubble defined. i.e. onMeasure
+    __callEventHandlerIfPresent(view, eventType, event);
+
+    if (event.bubbles)
+      __bubbleEvent(view._parent, eventType, event);
+  }
+}
+
+//@ts-ignore
+__BlueprintNative__.dispatchViewEvent = function dispatchEvent(viewId: string, eventType: string, event: any) {
+  if (__viewRegistry.hasOwnProperty(viewId)) {
+    const instance = __viewRegistry[viewId];
+
+    // Convert target/relatedTarget to concrete ViewInstance refs
+    if (event.target && __viewRegistry.hasOwnProperty(event.target)) {
+      event.target = __viewRegistry[event.target];
+    }
+
+    if (event.relatedTarget && __viewRegistry.hasOwnProperty(event.relatedTarget)) {
+      event.relatedTarget = __viewRegistry[event.relatedTarget];
+    }
+
+    // Convert native event object into it's SyntheticEvent equivalent if required.
+    if (SyntheticEvents.isMouseEventHandler(eventType))
+      event = new SyntheticMouseEvent(event);
+    else if (SyntheticEvents.isKeyboardEventHandler(eventType))
+      event = new SyntheticKeyboardEvent(event);
+
+    // If mouseDown event we store the target viewId as the last view
+    // to recieve a mouseDown for "onClick" book-keeping.
+    if (eventType === "onMouseDown") {
+      __lastMouseDownViewId = viewId;
+      __bubbleEvent(instance, eventType, event);
+      return;
+    }
+
+    if (eventType === "onMouseUp") {
+      __bubbleEvent(instance, eventType, event);
+
+      if (__lastMouseDownViewId && viewId === __lastMouseDownViewId) {
+        __bubbleEvent(instance, "onClick", event);
+      }
+
+      __lastMouseDownViewId = null;
```

Comment:
Hm... should this be inside the `if` block immediately preceding?

Reply:
Yeah arguably should be. Will tidy that up.

Reply:
I'm gonna go with your suggestion being right for the moment given the juce docs: 

`A mouseUp callback is sent to the component in which a button was pressed even if the mouse is actually over a different component when the button is released.`

Hopefully that holds in all cases otherwise there's a slight gotcha there if `__lastMouseDownViewId` isn't equal to `viewId`. I think it's correct in practice though. We just might need to keep an eye out if any weird edge case gets reported.